doc/adr/0001-jetstream-json-api-design.md

```diff
@@ -0,0 +1,138 @@
+# 1. JetStream JSON API Design
+
+Date: 2020-04-30
+
+## Status
+
+Proposed
+
+## Context
+
+At present, the API encoding consists of mixed text and JSON, we should improve consistency and error handling.
+
+## Decision
+
+We anticipate many types of consumer for JetStream including very small devices via MQTT and those written in 
+languages without easy to use JSON support like C.
+
+We will make it easy for light-weight clients by using `+OK` and `-ERR` style prefixes for client interactions,
+but the server management API will be JSON all the time. The rationale being that such small devices will most
+likely have their Consumers and Streams provisioned as part of the platform and often might not care for the
+additional context.
+
+In client responses we will use URL Encoding to provide additional context so multiple fields worth of
+context can be provided. URL Encoding has [broad language support](https://www.rosettacode.org/wiki/URL_decoding).
+
+### Admin APIs
+
+#### Requests
+
+All Admin APIs that today accept `nil` body should also accept an empty JSON document as request body.
+
+Any API that responds with JSON should also accept JSON, for example to delete a message by sequence we accept
+`10` as body today, this would need to become `{"seq": 10}` or similar.
```

Comment:
@derekcollison about this, I implemented a mix nil/byte/json body and its a bit awkward, I'd still like to suggest again that `{}` should be considered same as no body.

And updated here about taking in what comes out, so a API that returns JSON should accept JSON, looks like this only applies to the message delete for now.

Reply:
Agree on delete, will update as that is an admin/management API.

Reply:
For empty request, we could simply ignore the args, or force to be empty json `{}` but again what request structure in your program are you marshaling? That is where I am confused. I can add support for nil or `{}`.

Reply:
Resolved by https://github.com/nats-io/nats-server/commit/91cb1b051e27878ef79c7c06ab265829075c6009